src/vcpkg/base/unicode.cpp

```diff
@@ -3,176 +3,218 @@
+        if (code_unit < 0b1111'0000u)
+        {
+            if (3 > last - first)
             {
-                return {it, utf8_errc::InvalidCodeUnit};
+                out = end_of_file;
+                first = last;
+                return utf8_errc::UnexpectedEof;
             }
-            else if (kind != Utf8CodeUnitKind::Continue)
+
+            utf8_errc out_error;
+            if (bad_trailing(static_cast<unsigned char>(first[1]), out_error) ||
+                bad_trailing(static_cast<unsigned char>(first[2]), out_error))
             {
-                return {it, utf8_errc::UnexpectedStart};
+                out = end_of_file;
+                first = last;
+                return out_error;
             }
 
-            const int shift = 6 * (count - byte - 1);
-            code_point |= (code_unit & continue_mask) << shift;
+            // clang-format off
+            out = ((code_unit & 0b0000'1111) << 12)
+                | ((static_cast<unsigned char>(first[1]) & 0b0011'1111u) << 6)
+                |  (static_cast<unsigned char>(first[2]) & 0b0011'1111u);
+            // clang-format on
+            first += 3;
+            return utf8_errc::NoError;
         }
 
-        if (code_point > 0x10'FFFF)
+        if (code_unit < 0b1111'1000u)
         {
-            return {it, utf8_errc::InvalidCodePoint};
+            if (4 > last - first)
+            {
+                out = end_of_file;
+                first = last;
+                return utf8_errc::UnexpectedEof;
+            }
+
+            utf8_errc out_error;
+            if (bad_trailing(static_cast<unsigned char>(first[1]), out_error) ||
+                bad_trailing(static_cast<unsigned char>(first[2]), out_error) ||
+                bad_trailing(static_cast<unsigned char>(first[3]), out_error))
+            {
+                out = end_of_file;
+                first = last;
+                return out_error;
+            }
+
+            // clang-format off
+            out = ((code_unit & 0b0000'0111) << 18)
+                | ((static_cast<unsigned char>(first[1]) & 0b0011'1111u) << 12)
+                | ((static_cast<unsigned char>(first[2]) & 0b0011'1111u) << 6)
+                |  (static_cast<unsigned char>(first[3]) & 0b0011'1111u);
+            // clang-format on
+
+            if (out > 0x10'FFFF)
+            {
+                out = end_of_file;
+                first = last;
+                return utf8_errc::InvalidCodePoint;
+            }
+
+            first += 4;
+            return utf8_errc::NoError;
         }
 
-        out = code_point;
-        return {it, utf8_errc::NoError};
+        out = end_of_file;
+        first = last;
+        return utf8_errc::InvalidCodeUnit;
     }
 
     bool utf8_is_valid_string(const char* first, const char* last) noexcept
     {
-        utf8_errc err = utf8_errc::NoError;
-        for (auto dec = Utf8Decoder(first, last); dec != dec.end(); err = dec.next())
```

Comment:
Stealth bug fix: the previous implementation would crash if the first code point was invalid.